File: my_cpp_pkg/src/led_panel.cpp
#include "rclcpp/rclcpp.hpp"
#include <vector>
#include "my_robot_interfaces/srv/set_led.hpp"
#include "my_robot_interfaces/msg/led_state.hpp"


class LedPanelNode : public rclcpp::Node 
{
public:
    LedPanelNode() : Node("led_panel") 
    {
        this->declare_parameter("led_states", std::vector<int>({0, 0, 0}));
        this->led_state = this->get_parameter("led_states").as_integer_array();

        server_ = this->create_service<my_robot_interfaces::srv::SetLed>(
            "set_led",
            std::bind(&LedPanelNode::callbackSetLed, this,std::placeholders::_1, std::placeholders::_2));
        RCLCPP_INFO(this->get_logger(), "Led panel server has been started");

        publisher_ = this->create_publisher<my_robot_interfaces::msg::LedState>("led_panel_state",10);
        timer_ = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&LedPanelNode::publishLedState, this));
        RCLCPP_INFO(this->get_logger(), "Led state Publisher has been started.");
    }


    void callbackSetLed(const my_robot_interfaces::srv::SetLed::Request::SharedPtr request,
                            const my_robot_interfaces::srv::SetLed::Response::SharedPtr response)
    {
        if(request->led_number > 0 && request->led_number <= 3)
        {
            this->led_state[request->led_number - 1] = (request->state ? 1 : 0) ;
            response->success = true;
        }
        else
        {
            response->success = false;
        }

        RCLCPP_INFO(this->get_logger(), "Request led number %ld to %s resulted in %s", 
                    request->led_number,
                    (request->state ? "ON" : "OFF"),
                    (response->success ? "Success" : "Failure"));
    }

    void publishLedState()
    {
        auto msg = my_robot_interfaces::msg::LedState();
        msg.led_state[0] = this->led_state[0];
        msg.led_state[1] = this->led_state[1];
        msg.led_state[2] = this->led_state[2];
        publisher_->publish(msg);
    }

private:

    std::vector<int64_t> led_state;
    rclcpp::Service<my_robot_interfaces::srv::SetLed>::SharedPtr server_;

    rclcpp::Publisher<my_robot_interfaces::msg::LedState>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LedPanelNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/led_panel.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from my_robot_interfaces.srv import SetLed
from my_robot_interfaces.msg import LedState

class LedPanelNode(Node): 
    def __init__(self):
        super().__init__("led_panel") 

        self.declare_parameter("led_states", [0, 0, 0])

        self.led_state = self.get_parameter("led_states").value

        ## Create a server
        self.server_ = self.create_service(
            SetLed,"set_led", self.callback_set_led) # Srv_type, name_of_srv, callback
        self.get_logger().info("Set led server has been started")
        
        # Create publisher obj
        self.publisher_ = self.create_publisher(LedState, "led_panel_state", 10) # type, topic name, queue size

        # add timer
        self.timer_ = self.create_timer(0.5, self.publish_led_state)

        self.get_logger().info("Number Publisher has been started")

    def publish_led_state(self):
        # Create msg obj
        msg = LedState()
        # Add custom data
        for i in range(3):
            msg.led_state[i] = self.led_state[i]
        # Output the data
        self.publisher_.publish(msg)
        self.get_logger().info("Message published!")

    def callback_set_led(self, request, response):

        if(request.led_number > 0 and request.led_number <=3):
            self.led_state[request.led_number - 1] = 1 if (request.state) else 0
            response.success = True
        else:
            response.success = False
        
        self.get_logger().info(f"Request led number {str(request.led_number)} to {'ON' if request.state else 'OFF'} is {'Success' if response.success else 'Failure'}")
        return response     ## Always return the response

def main(args=None):
    rclpy.init(args=args)
    node = LedPanelNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()